### PR TITLE
CIF Related

### PR DESCRIPTION
**Pull request by MrSupW**
Complete the implementation of CIF-related code, including two types of Predictor and two types of CIF-Decoder. Meanwhile,  four baseline experiments were conducted on the AISHELL dataset.

### CODE REVIEW

File: wenet/cif/embedding.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
+# Copyright (c) 2023 ASLP@NWPU (authors: He Wang, Fan Yu)
```
Comment (robin1001) [inline wenet/cif/embedding.py:1]: The embedding.py shares almost the same code with transformer/embedding.py, I think we can reuse it.

File: wenet/cif/attention.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
+# Copyright (c) 2023 ASLP@NWPU (authors: He Wang, Fan Yu)
```
Comment (robin1001) [inline wenet/cif/attention.py:1]: same as embedding.py, attention shares almost the same code as transformer/attention.py, and transformer/attention.py takes streamming into account.

File: wenet/cif/utils.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+# FunASR(https://github.com/alibaba-damo-academy/FunASR)
+
+from typing import Optional
+
+import six
+import torch
+import numpy as np
+
+
+def make_pad_mask(lengths: torch.Tensor, length_dim: int = -1,
```
Comment (robin1001) [inline wenet/cif/utils.py:23]: this function is defined in `utils/mask.py`

File: wenet/cif/utils.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+    seq_range = torch.arange(0, maxlen, dtype=torch.int64)
+    seq_range_expand = seq_range.unsqueeze(0).expand(bs, maxlen)
+    seq_length_expand = lengths.unsqueeze(-1).to(seq_range_expand.device)
+    mask = seq_range_expand >= seq_length_expand
+
+    return mask
+
+
+def sequence_mask(lengths, maxlen: Optional[int] = None,
```
Comment (robin1001) [inline wenet/cif/utils.py:58]: this function is almost the same as subsequent_mask in utils/mask.py

**Review by robin1001 — APPROVED** [review]
